Cython/Includes/numpy/__init__.pxd

```diff
@@ -331,6 +333,8 @@ cdef extern from "numpy/arrayobject.h":
     ctypedef unsigned long long npy_uint96
     ctypedef unsigned long long npy_uint128
 
+    ctypedef npy_uint16   npy_half
```

Comment:
If one tried to do arithmetic with float16, would it simply interpret them as ints? It'd be better to prohibit it than allow this. Perhaps typedef this as an (empty) struct (named npy_half) instead.

Reply:
Hi, @robertwb . Thank your for your quick reply :-)

In tensorflow, we create a numpy array in Python, and we hope to convert it to Tensor. It's faster to hand it in C++ than in python, like the example below for uint16:

```python
def AppendFloat16ArrayToTensorProto(tensor_proto, np.ndarray[np.uint16_t, ndim=1] nparray):
      cdef long i, n
      n = nparray.size
      for i in range(n):
          tensor_proto.half_val.append(nparray[i])
```

However, because cython doesn't allow that we declare a `np.ndarray[np.float16_t, ndim=1] nparray`, we cannot do the same thing for `np.float16`.

Reply:
Yes, I agree that it's dangerous if someone try to do float16 operator (not supported in fact). I think we should hide it as you said, and don't allow people to do it. However I have no good idea, could you give me some suggestion? Thank you very much.

Reply:
Instead of `ctypedef npy_uint16   npy_half`, try

```
ctypedef struct npy_half:
    pass
```

(I'm not positive this will work, but think it should). This way you should be able to get and set `npy_half`s as values, and handle arrays of them, but not do arithmetic with them. (It won't change the definition under the hood, just Cython's interpretation if it.)